cli/src/pcluster/templates/cw_dashboard_builder.py

```diff
@@ -0,0 +1,517 @@
+        self.graph_width = 6
+        self.graph_height = 6
+        self.logs_width = 24
+        self.logs_height = 6
+        self.empty_section = True
+        self.dashboard = None
+
+        self._add_resources()
+
+    # -- Utility methods --------------------------------------------------------------------------------------------- #
+
+    @property
+    def _stack_region(self):
+        return core.Stack.of(self).region
+
+    # -- Resources --------------------------------------------------------------------------------------------------- #
+
+    def _add_resources(self):
+
+        # Initialize a cloudwatch dashboard
+        self.cloudwatch_dashboard = cloudwatch.Dashboard(
+            scope=self.stack_scope, id="CloudwatchDashboard", dashboard_name=self.dashboard_name
+        )
+
+        # Create a text widget for title "Head Node EC2 metrics"
+        self._add_text_widget("# Head Node EC2 Metrics")
+
+        # Head node instance graphs
+        self._add_head_node_instance_metrics_graphs()
+
+        VolumeMetric = namedtuple("VolumeMetric", ["title", "metrics"])
+        ebs_metrics = [
+            VolumeMetric("Read/Write Ops", ["VolumeReadOps", "VolumeWriteOps"]),
+            VolumeMetric("Read/Write Bytes", ["VolumeReadBytes", "VolumeWriteBytes"]),
+            VolumeMetric("Total Read/Write Time", ["VolumeTotalReadTime", "VolumeTotalWriteTime"]),
+            VolumeMetric("Queue Length", ["VolumeQueueLength"]),
+            VolumeMetric("Idle Time", ["VolumeIdleTime"]),
+        ]
+
+        ConditionalVolumeMetric = namedtuple("ConditionalVolumeMetric", ["title", "metrics", "supported_vol_types"])
+        conditional_ebs_metrics = [
+            ConditionalVolumeMetric("Consumed Read/Write Ops", "VolumeConsumedReadWriteOps", ["io1", "io2", "gp3"]),
+            ConditionalVolumeMetric("Throughput Percentage", "VolumeThroughputPercentage", ["io1", "io2", "gp3"]),
+            ConditionalVolumeMetric("Burst Balance", "BurstBalance", ["gp2", "st1", "sc1"]),
+        ]
+
+        # EBS and RAID metrics graphs
+        for storage_type, title in [
+            (SharedStorageType.EBS, "## EBS Metrics"),
+            (SharedStorageType.RAID, "## RAID Metrics"),
+        ]:
+            if len(self._storage_resource_ids[storage_type]) > 0:
+                self._add_volume_metrics_graphs(title, storage_type, ebs_metrics, conditional_ebs_metrics)
+
+        # EFS metrics graphs
+
+        if len(self._storage_resource_ids[SharedStorageType.EFS]) > 0:
+            self._add_efs_metrics_graphs()
+
+        # FSx metrics graphs
+        if len(self._storage_resource_ids[SharedStorageType.FSX]) > 0:
+            self._add_fsx_metrics_graphs()
+
+        # Head Node logs, if CW Logs are enabled
+        if self.config.is_cw_logging_enabled:
+            self._add_cw_log()
+
+    def _update_coord(self, d_x, d_y):
+        """Calculate coordinates for the new graph."""
+        self.coord._replace(x=self.coord.x + d_x)
```

Comment:
NamedTuple is immutable, so using _replace it's the only way to update it but it's not the cleanest way.
If we need to update coordinate values it's better to use a new Class for Coord, rather than a namedTuple.

Reply:
Done! Add a Coord class